Add atlas slice selection using similarity metrics

Elsword016 opened the pull request:
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
The current brainglobe-registration implementation relies on manually selecting specific 2D regions or 3D subvolume atlases for registration targets. This is time-consuming and potentially error-prone. This PR adds functionality to automatically identify optimal atlas slices for registration with sample brain images, significantly reducing manual effort and improving reproducibility.

**What does this PR do?**
This PR implements:
- Multiple similarity metrics for comparing images (NCC, MI, SSIM)
- Functions to compare sample images with entire atlas volumes
- Automatic selection of best-matching atlas slices
- Visualization tools for examining similarity across atlas slices

**Relation to GSOC project**
This PR is a foundational component for the "Automatic Atlas Region Selection" GSOC project. While the full GSOC project will expand to include advanced techniques like adaptive grid search, ML-based approaches, and Bayesian optimization for 3D subvolume selection, this PR implements:
- The core similarity measurement functionality needed for any region selection approach
- A framework for evaluating and comparing different selection methods

The initial implementation of automatic slice selection, which is the first step toward full 3D subvolume selection
By submitting this PR, I aim to demonstrate both my understanding of the registration challenges and my technical capability to implement solutions for the GSOC project. This implementation provides immediate benefits to users while establishing a solid foundation for the more comprehensive approaches to be developed during the GSOC project.
## References

This PR relates to the "Automatic Atlas Region Selection" GSOC project, which aims to eliminate manual selection of atlas regions in the registration workflow.

## How has this PR been tested?

The implementation has been tested with:
- The Allen Mouse Brain Atlas to validate compatibility with common use cases
- Quantitative comparison between different metrics to validate their effectiveness
- Demonstration of the full workflow in a Jupyter notebook
- Edge case testing with varying image sizes, empty slices, and potential NaN values

## Is this a breaking change?

No, this PR adds new functionality without modifying existing behaviour.

## Does this PR require an update to the documentation?

Yes, I've added comprehensive docstrings to all new functions. A demonstration notebook is also included to show usage examples. Further documentation updates could be added to reflect this new capability.

## Checklist:

- [x] The code has been tested locally
- [ ] Tests have been added to cover all new functionality (unit tests)
- [x] The documentation has been updated to reflect any changes
- [ ] The code has been formatted with [pre-commit](https://pre-commit.com/)

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,306 @@
+    """
+    # Ensure the images have the same shape
+    min_shape = np.minimum(image1.shape, image2.shape)
+    img1 = image1[: min_shape[0], : min_shape[1]].astype(np.float32)
+    img2 = image2[: min_shape[0], : min_shape[1]].astype(np.float32)
```
Why are we converting to float here?

IgorTatarnikov reviewed (changes requested):
This is a great start! I've update the `MANIFEST.in` file to exclude python notebooks, as they don't need to be shipped with the package. The tests look great, can you make sure they all pass? I added a small comment as to why the first test might be failing.

IgorTatarnikov commented:
Also, take a look at the tooling section of our [developer guide](https://brainglobe.info/community/developers/tooling.html). You can set up `pre-commit` hooks to run every time you commit to make sure the `linting` step of the test suite never fails on CI!

Elsword016 commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,306 @@
+    """
+    # Ensure the images have the same shape
+    min_shape = np.minimum(image1.shape, image2.shape)
+    img1 = image1[: min_shape[0], : min_shape[1]].astype(np.float32)
+    img2 = image2[: min_shape[0], : min_shape[1]].astype(np.float32)
```
if the inputs are in `uint8`, `uint16`, when doing the main calculations, numpy implicitly converts to `float64` to avoid overflow/underflow. But `float64` is compute heavy, so I just explicitly converted to `float32` to use less memory and also forcing all calculations into floating-point format, avoiding mixed-type issues.

Elsword016 commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,306 @@
+        Higher values indicate better similarity
+    """
+    # Normalize images to have same intensity range for histogram calculation
+    img1_norm = image1.astype(np.float32) / np.max(image1)
+    img2_norm = image2.astype(np.float32) / np.max(image2)
```
Same reason as before.

Elsword016 commented:
Thank you for your feedback and apologies for the delay, I will check all the issues, and push them as soon as possible.

Elsword016 commented:
I made the changes suggested, also getting into napari part a little bit, I added a widget to automatically move the z-slider to the slice predicted by the similarity metrics.  I think fixed the critical issues (imports, function signatures, etc.), pre-commit is still giving errors on the style, which I will address in a different PR

adamltyson commented:
Hi @Elsword016, would you mind fixing the linting issues in this PR? We require al checks to pass before merging a PR. Thanks!

Elsword016 commented:
okay i will do it

Elsword016 commented:
@adamltyson Fixed the linting issues it passes the pre-commit checks now. Please let me know if I need to make further changes
Thank you

IgorTatarnikov commented on .check-manifest-ignore:
This file shouldn't be part of this PR

IgorTatarnikov commented on MANIFEST.in:
```diff
@@ -3,6 +3,7 @@ include README.md
 include .napari-hub/DESCRIPTION.md
 include .napari-hub/config.yml
 include brainglobe_registration/napari.yaml
+include .check-manifest-ignore
```
We shouldn't ship the .check-manifest-ignore file with the package

IgorTatarnikov commented on MANIFEST.in:
```diff
@@ -12,3 +13,8 @@ recursive-exclude * __pycache__
 recursive-exclude docs *
 recursive-exclude tests *
+recursive-exclude notebooks *
+recursive-include test_dir *.csv
+recursive-include test_dir *.json
+recursive-include test_dir *.tiff
+recursive-include test_dir *.txt
```
```suggestion
```

`test_dir` shouldn't be part of this PR. Regression test data already exists in `tests/test_images`.

IgorTatarnikov reviewed (changes requested):
Hey, @Elsword016 it looks like you added some extra files as part of this PR. Would you mind removing them before I do a full review? There's no need to include `test_dir` as regression test data already exists in `tests/test_images`.

Could I also ask that we split this PR into two, one for implementing and testing the similarity metrics, and one for adding it as a UI element?

IgorTatarnikov reviewed (changes requested):
Nearly there! There are still 3 tests that are failing. You can check which ones by looking at the results of the [CI run](https://github.com/brainglobe/brainglobe-registration/actions/runs/14097394863/job/39505721554?pr=82). You can also run the tests locally with `pytest tests/test_similarity_metrics.py` from the root repo directory given that you've installed the optional `dev` dependencies when setting up your dev environment.

Elsword016 commented:
Changes:
1. Removed `elastix` references (working on it though in a separate branch)
2. Added `np.nan_to_num()` to handle NaN values in inputs across the metrics.
3. Checks for zero division.
4. Explicit checks for NaN results and return 0.0 instead
5. For `mutual_information()` function:
I. I check if both images are constant instead of if either is constant.
II.  If both are constant AND have the same value, I return 1.0 (positive MI for identical images)
III. If both are constant but have different values or if only one is constant, I return 0.0.
I think this may not be the best way to handle this, though it passed `pytest` checks. I would appreciate any suggestions you might have.
6. Removed the extra validation from the `test_similarity_metrics.py` as suggested before.

Elsword016 commented:
@IgorTatarnikov  take your time, I need to push some more fixes

Elsword016 commented:
I don't have a macOS build to check, but some sort of `vispy` error from the logs. The windows test is working correctly when running `tox` locally. Can you check?

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,334 @@
+
+    try:
+        # Normalize images to [0,1] for histogram calculation
+        if np.max(img1_crop) != 0:
+            img1_norm = img1_crop.astype(np.float32) / np.max(img1_crop)
+        else:
+            img1_norm = img1_crop.astype(np.float32)
+
+        if np.max(img2_crop) != 0:
+            img2_norm = img2_crop.astype(np.float32) / np.max(img2_crop)
+        else:
+            img2_norm = img2_crop.astype(np.float32)
+
```
If you're checking whether the images are already normalised shouldn't this be `np.max(img) != 1.0`?

Otherwise, the images we're working with are typically stored as unsigned int and shouldn't have any negative values.

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,334 @@
+            cropped_moving = moving_image[:min_height, :min_width]
+            cropped_atlas = atlas_slice[:min_height, :min_width]
+
+            similarity = metric_func(cropped_atlas, cropped_moving)
+            results[slice_idx] = similarity
```
```suggestion
            results[slice_idx] = metric_func(cropped_atlas, cropped_moving)
```

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,334 @@
+    if search_range is None:
+        search_range = (0, atlas_volume.shape[0])
+
+    # Create a range with the specified step size
+    slice_indices = range(search_range[0], search_range[1], search_step)
+
+    # Compare the image with the specified slices
+    similarities = compare_image_to_atlas_slices(
+        moving_image,
+        atlas_volume,
+        slice_range=(slice_indices[0], slice_indices[-1] + 1),
+        metric=metric,
+    )
```
You create a range here to allow `search_step` to be used. However, this information isn't passed back to the `compare_image_to_atlas_slices` method.

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,359 @@
+            try:
+                self.test_image = io.imread(str(sample_hipp_path))
+
+                # If image is 3D or RGB, handle appropriately
+                if len(self.test_image.shape) > 2:
+                    if len(
+                        self.test_image.shape
+                    ) == 3 and self.test_image.shape[2] in [3, 4]:
+                        self.test_image = np.mean(
+                            self.test_image[:, :, :3], axis=2
+                        )
+                    else:
+                        self.test_image = self.test_image[0]
```
The test image is always going to be a one channel tiff. You don't need the extra logic to validate the input here.

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,359 @@
+        sample_hipp_path = test_images_dir / "sample_hipp.tif"
+        atlas_hipp_path = test_images_dir / "Atlas_Hipp.tif"
+
+        # Load sample image if it exists
+        if sample_hipp_path.exists():
+            try:
```
I don't think you need to test whether `sample_hipp_path` exists. The automated testing suite will always clone the full repository and therefore will have access to the `test_images` folder, developers should also be working out of a local clone of the repo and have access to the test_images folder as well.

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,359 @@
+        else:
+            self.test_image = self.img_random
+
+        # Load atlas image if it exists
+        if atlas_hipp_path.exists():
+            try:
+                atlas_image = io.imread(str(atlas_hipp_path))
+
+                # Create mini atlas from the 2D image
+                self.atlas_volume = np.stack([atlas_image for _ in range(5)])
+            except Exception as e:
+                print(f"Could not load Atlas_Hipp.tif: {e}")
+                # Fallback to synthetic atlas (already created)
```
Same as above in terms of `atlas_hipp_path` existing. You can also load a `BrainGlobeAtlas` object here for testing. The test suite sets up `allen_mouse_25um` to be one of the available atlases. This can be used as a target volume.

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,359 @@
+
+    def test_mutual_information_different(self):
+        """Test MI with different images."""
+        result = mutual_information(self.img_random, self.img_gradient)
+        assert isinstance(result, float)
```
This test should check for a value to be strictly larger than the minimum returned by `skimage.metrics.mutual_information` which is 1.0.

Right now this runs through the except block of the `try except` in `mutual_information` making the user think the function ran. It might be better to print the exception for the user, rather than silently handling it.

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,334 @@
+            return 0.0
+
+        return ssim_value
+    except Exception:
```
It's not a good idea to have a catch-all `except`. This silently disregards every possible error without notifying the user.

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,359 @@
+        # Two identical constant images have perfect MI
+        assert result == 1.0
+
+        # Test with very small images that might cause calculation issues
+        tiny_img1 = np.random.rand(2, 2)
+        tiny_img2 = np.random.rand(2, 2)
+        result = mutual_information(tiny_img1, tiny_img2)
+        # Should still return a float value
+        assert isinstance(result, float)
```
I don't think this test is necessary

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,359 @@
+        result = structural_similarity_index(img_zeros, img_zeros)
+        assert not np.isnan(result)
+
+        # Test with very small images that might cause calculation issues
+        tiny_img1 = np.random.rand(2, 2)
+        tiny_img2 = np.random.rand(2, 2)
+        result = structural_similarity_index(tiny_img1, tiny_img2)
+        # Should still return a float value
+        assert isinstance(result, float)
```
Same as above

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,359 @@
+
+    def test_find_best_atlas_slice(self):
+        """Test the find_best_atlas_slice function."""
+        from brainglobe_registration.similarity_metrics import (
+            compare_image_to_atlas_slices,
+        )
```
Why is this imported twice?

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,359 @@
+        # Should still return a float value
+        assert isinstance(result, float)
+
+    def test_find_best_atlas_slice(self):
```
I think this test needs to be split up into multiple smaller test.

IgorTatarnikov reviewed (changes requested):
I've left some comments, this is on the right track but still needs a bit of work.

Elsword016 commented:
Based on the feedback, I made a lot of changes based on that and also took some time to improve clarity, removing unnecessary logic and other mistakes:

- Fixed imports (wrong, duplicates, etc)
- Created a centralized NaN handling in `_match_image_sizes()` and return a clear status flag, removing repetitive NaN checking code from individual metric functions
- Added a dedicated `_normalize_image()` helper function.
- Created `METRIC_FUNCTIONS` dictionary to map metric names to functions, removing if, elif chains
- Used lambda functions to ensure metric functions are properly mocked in tests
- Added proper return types for all the functions.
- Replaced catch-all exceptions with specific exception types in all the functions.
- Used the `mocker.patch()` in the similarity_metrics test for patching the functions.

Please let me know if I have make further changes.

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,372 @@
+    int
+        Index of the best matching slice
+    """
+    if search_range is None:
+        slice_indices = range(0, atlas_volume.shape[0], search_step)
+    else:
+        start_slice, end_slice = search_range
+        # Ensure start/end are within bounds
+        start_slice = max(0, start_slice)
+        end_slice = min(atlas_volume.shape[0], end_slice)
+        slice_indices = range(start_slice, end_slice, search_step)
```
I think this should be moved into the `compare_image_to_atlas_slices` function.

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,372 @@
+    # Check if similarity_results is empty (e.g., invalid range)
+    if not similarity_results:
+        warnings.warn("No valid slices found in the specified search range.")
+        # Return first index of the search range as fallback
+        return slice_indices[0] if len(slice_indices) > 0 else 0
```
I would expect this to either return an invalid index, or an exception. Returning a valid index may be confusing for downstream applications.

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,372 @@
+    # had NaNs)
+    if similarity_results[best_slice_index] == float("-inf"):
+        warnings.warn("All compared slices yielded invalid similarity values.")
+        # Return first index of the search range as fallback
+        return slice_indices[0] if len(slice_indices) > 0 else 0
```
Again, I don't think this should return a valid index if all comparisons failed.

IgorTatarnikov reviewed (changes requested):
I've added some inline comments. 

The biggest thing that needs to be fixed relates to the results of the `mutual_information` calculations. It should return `2` for perfectly correlated images, and `1` for uncorrelated images. 

I also think we should be throwing `Exceptions` in several places where you've raised warnings. Depending on the amount of text that's output to the terminal, or logged to a file it might be harder to find the warnings when debugging. It's okay to throw informative exceptions when encountering certain invalid states. This gives immediate feedback to the user, otherwise they may not know something went wrong until the entire pipeline finishes.

Elsword016 commented:
Hi @IgorTatarnikov i will look at comments and make changes. Apologies for the delay, I had some medical emergencies

Elsword016 commented:
Hi @IgorTatarnikov, I made the changes according to your suggestions. Looking forward towards your feedback.

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -119,23 +108,19 @@ def mutual_information(
-        return 1.0 if img1_proc.flat[0] == img2_proc.flat[0] else 0.0
+        return 2.0 if img1_proc.flat[0] == img2_proc.flat[0] else 1.0
     elif is_img1_constant or is_img2_constant:
         # If only one image is constant, MI is 0
```
```suggestion
        # If only one image is constant, MI is 1
```

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -195,52 +174,24 @@ def structural_similarity_index(img1: npt.NDArray, img2: npt.NDArray) -> float:
 
         return ssim_value
     except (ValueError, RuntimeError, TypeError) as e:
         # If any error occurs during calculation, print warning and return -inf
```
```suggestion
        # If any error occurs during calculation inform user and exit
```

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,322 @@
+
+        return mi_value
+    except (ValueError, RuntimeError, TypeError) as e:
+        # If any error occurs during calculation, print warning and return -inf
```
```suggestion
        # If any error occurs during calculation inform user and exit
```

IgorTatarnikov commented on brainglobe_registration/similarity_metrics.py:
```diff
@@ -0,0 +1,322 @@
+        key=lambda k: similarity_results[k],
+    )
+
+    if similarity_results[best_slice_index] == float("-inf"):
+        raise ValueError(
+            "All compared slices yielded invalid similarity values."
+        )
```
I believe this is unnecessary now as none of the metrics can return -inf

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,367 @@
+        ), f"Expected slice 1 within range {search_range}, got {best_slice}"
+
+    def test_find_best_slice_all_invalid(self):
+        """Test finding best slice when all comparisons are invalid (-inf)."""
```
Need to update this docstring

IgorTatarnikov commented on tests/test_similarity_metrics.py:
```diff
@@ -0,0 +1,367 @@
+            side_effect=ValueError("Deliberate test exception"),
+        )
+
+        # This should catch the exception from the mocked NCC and return -inf
```
```suggestion
        # This should catch and propagate the exception from the mocked NCC
```

IgorTatarnikov reviewed (changes requested):
Looks great! Just a few more nitpicky comments about the documentation, apart from that I think it's ready to go!

Also, we should add `skimage` as a dependency in `pyproject.toml` as we now directly depend on functions from the package.

Elsword016 commented:
I made the changes that were suggested.

Elsword016 commented:
there are some failures in the `test_registration_widget.py` , would you mind having a look??

adamltyson commented:
At least one of the test failures seems related to https://github.com/brainglobe/brainglobe-registration/pull/88. Could you try merging main to see if this fixes anything?

Elsword016 commented:
Okay, I will try pushing again. There's a problem with the Python version. Attaching two screenshots. pytest passes in `python 3.11.11` but fails with `python 3.13.2`
![Python 3.11](https://github.com/user-attachments/assets/d4d3b38d-3223-470a-9bc4-523c89b95b2c)

![Python 3.13](https://github.com/user-attachments/assets/f1f1dbbe-4c5f-4287-b4cb-ee75744c5ff5)

Elsword016 commented:
Hi @adamltyson can you try running the actions again. I manually had to fix the `pyproject.toml` with the napari version for some reason `git merge origin/main` is not working (not fetching the newer file) even though the commit hashes are same.

IgorTatarnikov commented on pyproject.toml:
```diff
@@ -34,8 +34,8 @@ dependencies = [
+    "qt-niu",
     "scikit-image",
-    "scipy",
     "qt-niu"
```
I think `scipy` somehow disappeared from the requirements list. Could it be added back? 

```suggestion
    "scipy"
```

IgorTatarnikov reviewed (approved):
Just a tiny comment about the changes in `pyproject.toml`. By moving `qt-niu` into alphabetical order (thank you!)  you got rid of `scipy` and duplicated `qt-niu`. Apart from that it's good to go!

Thanks a lot for your contribution! This is fantastic.
